timeline_sync/api.py

```diff
@@ -0,0 +1,172 @@
+
+    updates = [user_timeline_item.to_json() for user_timeline_item in user_timeline]
+
+    user_timeline.delete()
+    db.session.commit()
+
+    result = {
+        "updates": updates,
+        "syncURL": url_for('api.sync', _external=True)
+    }
+    return jsonify(result)
+
+
+@api.route('/user/pins/<pin_id>', methods=['PUT', 'DELETE'])
+def user_pin(pin_id):
+    try:
+        user_token = request.headers.get('X-User-Token')
+        user_id, app_uuid, data_source = parse_user_token(user_token)
+    except ValueError:
+        return api_error(410)
+
+    if request.method == 'PUT':
+        pin_json = request.json
+        if pin_json is None or pin_json.get('id') != pin_id:
+            return api_error(400)
+
+        pin = TimelinePin.query.filter_by(app_uuid=app_uuid, user_id=user_id, id=pin_id).one_or_none()
+        if pin is None:  # create pin
+            pin = TimelinePin.from_json(pin_json, app_uuid, user_id, data_source, 'web', [])
+            if pin is None:
+                return api_error(400)
+
+            user_timeline = UserTimeline(user_id=user_id,
+                                         type='timeline.pin.create',
+                                         pin=pin)
+            db.session.add(pin)
+            db.session.add(user_timeline)
+            db.session.commit()
+        else:  # update pin
+            try:
+                pin.update_from_json(pin_json)
+                user_timeline = UserTimeline(user_id=user_id,
+                                             type='timeline.pin.create',
+                                             pin=pin)
+                db.session.add(pin)
+                db.session.add(user_timeline)
+                db.session.commit()
+            except (KeyError, ValueError):
+                return api_error(400)
+
+    elif request.method == 'DELETE':
+        pin = TimelinePin.query.filter_by(app_uuid=app_uuid, user_id=user_id, id=pin_id).first_or_404()
+        user_timeline = UserTimeline(user_id=user_id,
+                                     type='timeline.pin.delete',
+                                     pin=pin)
+        db.session.add(user_timeline)
+        db.session.commit()
+    return 'OK'
+
+
+@api.route('/shared/pins/<pin_id>', methods=['PUT', 'DELETE'])
+def shared_pin(pin_id):
+    api_key = request.headers.get('X-API-Key')
+    return api_error(403)  # TODO: check api_key in dev portal, add/delete shared pins
+
+
+@api.route('/user/subscriptions')
+def user_subscriptions_list():
+    try:
+        user_token = request.headers.get('X-User-Token')
+        user_id, app_uuid, _ = parse_user_token(user_token)
+    except ValueError:
+        return api_error(410)
+
+    return jsonify(
+        {
+            "topics": [sub.topic for sub in UserSubscription.query.filter_by(app_uuid=app_uuid, user_id=user_id)]
+        }
+    )
+
+
+@api.route('/user/subscriptions/<topic>', methods=['POST', 'DELETE'])
```

Comment:
Given that shared pins don't work, I wonder whether it would be worth making this fail, too.